substrate/frame/asset-conversion/src/lib.rs

```diff
@@ -765,22 +756,119 @@ pub mod pallet {
+		/// is provided and the swap can't achieve at least this amount, an error is returned.
+		///
+		/// On a successful swap, the function returns the `credit_out` of `path[last]` obtained
+		/// from the `credit_in`. On failure, it returns an `Err` containing the original
+		/// `credit_in` and the associated error code.
+		///
+		/// WARNING: This may return an error after a partial storage mutation. It should be used
+		/// only inside a transactional storage context and an Err result must imply a storage
+		/// rollback.
+		pub(crate) fn do_swap_exact_credit_tokens_for_tokens(
+			path: BoundedVec<T::MultiAssetId, T::MaxSwapPathLength>,
+			credit_in: Credit<T>,
+			amount_out_min: Option<T::AssetBalance>,
+		) -> Result<Credit<T>, (Credit<T>, DispatchError)> {
+			let amount_in = match credit_in.peek() {
+				Ok(a) => a,
+				Err(e) => return Err((credit_in, e)),
+			};
+			let inspect_path = |credit_asset| {
+				ensure!(path.get(0).map_or(false, |a| *a == credit_asset), Error::<T>::InvalidPath);
+				ensure!(!amount_in.is_zero(), Error::<T>::ZeroAmount);
+				ensure!(amount_out_min.map_or(true, |a| !a.is_zero()), Error::<T>::ZeroAmount);
+
+				Self::validate_swap_path(&path)?;
+				let path = Self::balance_path_from_amount_in(amount_in, path)?;
+
+				let amount_out = path.last().map(|(_, a)| *a).ok_or(Error::<T>::InvalidPath)?;
+				ensure!(
+					amount_out_min.map_or(true, |a| amount_out >= a),
+					Error::<T>::ProvidedMinimumNotSufficientForSwap
+				);
+				Ok(path)
+			};
+			let path = match inspect_path(credit_in.asset()) {
+				Ok(p) => p,
+				Err(e) => return Err((credit_in, e)),
+			};
+
+			let credit_out = Self::credit_swap(credit_in, &path)?;
+
+			Self::deposit_event(Event::CreditSwapExecuted { path });
+
+			Ok(credit_out)
+		}
+
+		/// Swaps a portion of `credit_in` of `path[0]` asset to obtain the desired `amount_out` of
+		/// the `path[last]` asset. The provided `credit_in` must be adequate to achieve the target
+		/// `amount_out`, or an error will occur.
+		///
+		/// On success, the function returns a (`credit_out`, `credit_change`) tuple, where
+		/// `credit_out` represents the acquired amount of the `path[last]` asset, and
+		/// `credit_change` is the remaining portion from the `credit_in`. On failure, an `Err` with
+		/// the initial `credit_in` and error code is returned.
+		///
+		/// WARNING: This may return an error after a partial storage mutation. It should be used
+		/// only inside a transactional storage context and an Err result must imply a storage
+		/// rollback.
+		pub(crate) fn do_swap_credit_tokens_for_exact_tokens(
+			path: BoundedVec<T::MultiAssetId, T::MaxSwapPathLength>,
+			credit_in: Credit<T>,
+			amount_out: T::AssetBalance,
+		) -> Result<(Credit<T>, Credit<T>), (Credit<T>, DispatchError)> {
+			let amount_in_max = match credit_in.peek() {
+				Ok(a) => a,
+				Err(e) => return Err((credit_in, e)),
+			};
+			let inspect_path = |credit_asset| {
```

Comment:
helps to avoid matching a non clone credit object for returning it in case of error 
alternative solution could be - https://gist.github.com/muharem/24376bcaa4ab19c3e7e0974cbc063648